### DISCUSSION TITLE
Python / pep440 & how Renovate handles a dependency without version specified

### DISCUSSION THREAD

**Discussion by stibi** [renovatebot/renovate#19358] [Request Help] [answered]
Hello,
so I have a project, with a couple of requirements.txt files, for various purposes. There is a bunch of dependencies, some with a version specified, but most of the not. So, whatever pip gets at the build time, that goes into the package. Not a good plan.

So I'm configuring Renovate to help me with this and I'm curious how to configure it, so it does start taking care about all those dependencies without any version specified. Because it ignores them, I can see in the log:

```
             "packageFile": "mysql/maintenance/requirements.txt",
             "deps": [
               {
                 "depName": "PyMySQL",
                 "datasource": "pypi",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [],
                 "versioning": "pep440",
                 "skipReason": "invalid-value"
               },
```

Of course, I can do that myself, manually, that's what I'll do if there is no other way. Can Renovate do it for me? What am I missing?

The config I'm using is pretty basic. I was trying a bit to push Renovate into what I'm describing here, hence the explicit `rangeSrategy: pin` there, but that and few others were just pretty blind attempts.

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base",
    ":separateMajorReleases",
    ":separateMultipleMajorReleases",
    ":label(dependencies)"
  ],
  "platform": "gitlab",
  "pip_requirements": {
    "fileMatch": ["^requirements-?.*\\.txt"]
  },
  "packageRules": [
    {
      "matchPackagePatterns": ["*"],
      "rangeStrategy": "pin"
    },
    {
      "matchUpdateTypes": ["minor", "patch"],
      "automerge": true
    },
    {
      "matchUpdateTypes": [
        "major"
      ],
      "labels": [
        "major"
      ]
    },
    {
      "matchUpdateTypes": [
        "minor"
      ],
      "labels": [
        "minor"
      ]
    },
    {
      "matchUpdateTypes": [
        "patch"
      ],
      "labels": [
        "patch"
      ]
    }
  ]
}
```

Thanks a lot for any hint or simply clarifying that Renovate can't do that for me, that's totally valid, too.

**Comment by rarkins** [comment, marked as answer]
Right now Renovate can only pin if there's a "real" constraint like "1.x" and does not have the capability to pin if the range is completely empty.

This is simply a technical limitation and nothing philosophical. Ironically it's harder to replace "nothing with something" than it is to replace "something with something else"
